### ISSUE TITLE
Unexpected performance drop [Corpse Party]

### ISSUE THREAD

**Issue by Nabokov86** [hrydgard/ppsspp#15348]
### Game or games this happens in

ULJM05704 - コープスパーティー ブラッドカバー リピーティッドフィアー

### What area of the game

The game runs very fast, but sometimes there is a big drop in performance. Especially when you go from one location to another.
This causes the sound to stutter for a moment, which is annoying.

### Speed seen in PPSSPP

With unlimited speed - 800%, when dropped - 30-60%.

### GE frame capture and debug statistics

![ULJM05704_00006](https://user-images.githubusercontent.com/26260168/150990172-8dcdc5c6-f8d6-41f4-ba7d-934be879d429.jpg)

### Platform

Android

### Mobile phone model or graphics card

Mali-G31 GPU

### PPSSPP version affected

v1.12.3-776-g1c18c172a

### Graphics backend (3D API)

OpenGL / GLES

### Any other notes or things you've tried

I tried resetting the settings, tried other android devices, same issue.

### Checklist

- [X] Test in the [latest git build](https://buildbot.orphis.net/ppsspp/) in case it's already fixed.
- [X] [Search for other reports](https://github.com/hrydgard/ppsspp/search?q=my+issue&type=issues) of the same issue.
- [X] Try resetting settings or older versions and include if the issue is related.
- [X] Try changing graphics settings to determine if one causes the slowdown.
- [ ] Include logs or screenshots of issue.

**Comment by Nabokov86** [comment]
These peaks occur every time you enter a different area.
![ULJM05704_00011](https://user-images.githubusercontent.com/26260168/151133788-1db9d614-bb5e-4e2c-9393-50ac25a42676.jpg)

**Comment by hrydgard** [comment]
Hm, those peaks look quite nasty indeed..

**Comment by unknownbrackets** [comment]
Do the same peaks, just with a presumably lower amplitude, occur on PC?  I want to make sure this isn't IO related - I know Android recently made IO much slower.

It's interesting that LockLwMutex is so hot, it's probably some sort of tight loop, I guess.  I don't think that has timings yet.

-[Unknown]

**Comment by Nabokov86** [comment]
@unknownbrackets The desktop computer handles it better, but I think the peaks are relatively the same.
![ULJM05704_00002](https://user-images.githubusercontent.com/26260168/151695588-f4e37580-e088-4cd5-a4f7-2e5d5a2722e7.jpg)

And with unlimited speed:
![ULJM05704_00001](https://user-images.githubusercontent.com/26260168/151695618-83f88083-979d-41fd-afbf-c240cd1970b6.jpg)

**Comment by hrydgard** [comment]
@Nabokov86 Does it work better in the latest builds (829, for example?)

**Comment by Nabokov86** [comment]
@hrydgard It looks the same, no difference

**Comment by Nabokov86** [comment]
Look at this!

Android:
![ULJM05704_00013](https://user-images.githubusercontent.com/26260168/151696576-ae996f38-eccc-4d87-a80f-7eeadf8c6510.jpg)

**Comment by hrydgard** [comment]
Still not great, but looks better indeed! Do you still get audio glitches?

 I wonder if this happens to any degree on the real PSP, or if there are more related timing issues somehow.

**Comment by Nabokov86** [comment]
@hrydgard Audio glitches are still present, but they are shorter and much less annoying. 

Still sometimes the sound is noticeably interrupted, especially in certain areas.

**Comment by unknownbrackets** [comment]
Do the debug statistics show anything different now?

-[Unknown]

**Comment by Nabokov86** [comment]
@unknownbrackets No :(
![ULJM05704_00035_1](https://user-images.githubusercontent.com/26260168/151708265-b521d53e-64d6-4360-a07a-e4c491113061.jpg)

**Comment by unknownbrackets** [comment]
Strange.  I'd be less surprised if sceAtracDecodeData was maybe the cuplrit, but I wonder why it's locking mutexes enough that it's noticeably slow... I wonder if unlock is just behind it in slowness...?

When #15368 is merged, it may be easier to catch these blips using frame advance, and especially see if there's interesting stats right before or right after.  I'm also interested in whether the time taken is significantly different when frame advancing.

-[Unknown]

**Comment by Nabokov86** [comment]
@unknownbrackets I apologize for the late response. 

#15368 is merged, right?
I can't find the frame advance option. Could you please help me with that?

**Comment by unknownbrackets** [comment]
You can map a button in control mapping to frame advance.  Then when you press that button, it will pause, and you can keep pressing it to step forward one frame at a time.  To stop frame advance, press whatever button is mapped to fast-forward.

-[Unknown]

**Comment by Nabokov86** [comment]
@unknownbrackets  Thanks for the quick reply!

Checked twice, I definitely don't have that button. Could this option be unavailable on android?
I will try it on Windows later.

**Comment by Nabokov86** [comment]
@unknownbrackets  Yes, frame advance is not available on android.

I recorded two videos: using frame advance and 5% speed. The result is quite different.

5% speed:
https://user-images.githubusercontent.com/26260168/152365423-6d69079d-c202-4f1a-9dcf-c82da5a703d5.mp4
Here you can clearly see the lag and LockLwMutexCB.

Frame advance:
https://user-images.githubusercontent.com/26260168/152367211-570ae50b-dbf8-442e-a191-459d4eb96cf5.mp4

Hope this will be useful.

**Comment by Nabokov86** [comment]
Android:

https://user-images.githubusercontent.com/26260168/152378315-1a9bb0d7-e2b7-4c3f-b4a0-7e93b047e68d.mp4

**Comment by unknownbrackets** [comment]
Has anything improved in the very latest git builds?  We made another change, which might help.

-[Unknown]

**Comment by Nabokov86** [comment]
@unknownbrackets No, nothing.

**Comment by unknownbrackets** [comment]
There was another recent change, has this improved on Android?

-[Unknown]

**Comment by Nabokov86** [comment]
@unknownbrackets Yes! I don't hear audio glitches anymore. It sounds very smooth.

However, peaks on the graph look the same. A very short image lag is noticeable during peaks as before. Despite this, everything looks very smooth to me :rocket:
![ULJM05704_00059](https://user-images.githubusercontent.com/26260168/162489646-cbd61522-3078-4fa7-bad0-2d23699955d0.jpg)

**Comment by Nabokov86** [comment]
In other locations there are even bigger peaks, but the sound is still surprisingly smooth.
![ULJM05704_00053](https://user-images.githubusercontent.com/26260168/162490558-6ff19c76-aa72-4615-9d2e-926985b3e7f3.jpg)
![ULJM05704_00057](https://user-images.githubusercontent.com/26260168/162490566-378bcfe2-37a0-46fb-b234-0fa6458610bc.jpg)

**Comment by Nabokov86** [comment]
I tested it now on a slower old android device, the sound is interrupted during the location change as before. 

I also think maybe I changed something in the settings or some other change in the device, so the audio glitches disappeared. Or it's an improvement due to the recent changes in PPSSPP. It's hard to tell, I'm not sure now. 
If it is important I can test the previous builds again to be sure. @unknownbrackets I'm sorry, сould you tell me the build before these changes?

**Comment by Nabokov86** [comment]
I discovered a new thing. Bluetooth-frendly buffer reduces audio glitches a lot in this game.

**Comment by hrydgard** [comment]
Yeah well that means that we still have stalls, just that the buffer is large enough with that setting to cover for it..

**Comment by unknownbrackets** [comment]
I have been thinking of putting atrac decode on a thread, which might help this.  I suspect the game is decoding the entire background music for the next area during the scene transition.

-[Unknown]

**Comment by Nabokov86** [comment]
> I suspect the game is decoding the entire background music for the next area during the scene transition

I don't think so. The background music does not change during scene transitions.

For example, there is a dead end in this scene. The background music does not change, there seems to be no new sounds at all. The only new content is textures and a text note. The textures are loaded before this peak on the graph. Every time you enter this location causes this. Looks like something is loaded/decoded every time for this particular scene.

![ULJM05704_00051_1](https://user-images.githubusercontent.com/26260168/162584467-62582bc7-39ba-4318-acda-0e6c02369424.jpg)

**Comment by EternalsDarkhole** [comment]
Cache full ISO in RAM doesn't help this if you enable?
![IMG_20220411_114848](https://user-images.githubusercontent.com/37603562/162662247-7af70f6f-4427-41aa-81c1-ec0430c82d3b.jpg)

**Comment by EternalsDarkhole** [comment]
Can you upload save state of this area/stage of this game where it lag.

**Comment by Nabokov86** [comment]
@EternalsDarkhole Sure.

1) [ULJM05704_1.03_0.zip](https://github.com/hrydgard/ppsspp/files/8481243/ULJM05704_1.03_0.zip) Go up here.
2) [ULJM05704_1.03_1.zip](https://github.com/hrydgard/ppsspp/files/8481251/ULJM05704_1.03_1.zip) Go left here.

It lags almost everywhere more or less, though.

**Comment by EternalsDarkhole** [comment]
> @EternalsDarkhole Sure.
> 
> 1. [ULJM05704_1.03_0.zip](https://github.com/hrydgard/ppsspp/files/8481243/ULJM05704_1.03_0.zip) Go up here.
> 2. [ULJM05704_1.03_1.zip](https://github.com/hrydgard/ppsspp/files/8481251/ULJM05704_1.03_1.zip) Go left here.
> 
> It lags almost everywhere more or less, though.

Thanks!
BTW is NPUH-10117 is the usa version?

**Comment by Nabokov86** [comment]
> NPUH-10117 is the usa version

Yeah, I think so. But I have the Japanese version ULJM-05704

**Comment by Gamemulatorer** [comment]
> Android:
> 
>  Android_debug.mp4

I cannot reproduce this kind of issue on the usa version of this game on my phone using ppsspp latest build.

https://user-images.githubusercontent.com/37603562/163184231-ef382dca-23f1-45fa-a067-f718d359744b.mp4

**Comment by Gamemulatorer** [comment]
Maybe this issue is similar to KoF Orochi Collection https://github.com/hrydgard/ppsspp/issues/12543 where the USA version is smooth while the JAP version is slow.

**Comment by Gamemulatorer** [comment]
> Yes, I confirm. Looks like it's only related to the Japanese version.
> 
> NPUH-10117: ![NPUH10117_00003_1](https://user-images.githubusercontent.com/26260168/163197053-58973626-a406-407d-8541-7fd0d9197790.jpg)

Can you try software rendering?
